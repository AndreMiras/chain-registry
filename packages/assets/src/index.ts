import asset_lists from './asset_lists';

export default {
  asset_lists
};

export { asset_lists };
