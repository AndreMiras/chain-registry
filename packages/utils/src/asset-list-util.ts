import { Asset, AssetList, AssetDenomUnit } from '@chain-registry/types';
import BigNumber from 'bignumber.js';

export type CoinDenom = AssetDenomUnit['denom'];

export type Exponent = AssetDenomUnit['exponent'];

export interface CoinGeckoUSD {
  usd: number;
}

export interface PriceHash {
  [key: CoinDenom]: number;
}

export function getAssetByDenom(assets: Asset[], denom: CoinDenom): Asset {
  const asset = assets.find((asset) => asset.base === denom);
  if (!asset) {
    throw new Error(`Asset not found: ${denom}`);
  }
  return asset;
}

export function getDenomByCoinGeckoId(
  assets: Asset[],
  coinGeckoId: string
): CoinDenom {
  return assets.find((asset) => asset.coingecko_id === coinGeckoId).base;
}

type GetCoinGeckoIdByDenomOptions = {
  allowTestnet?: boolean;
  customAssetFilter?: (asset: Asset) => boolean;
  excludedChainNames?: string[];
};

export function getCoinGeckoIdByDenom(
  assets: AssetList[],
  denom: CoinDenom,
  {
    allowTestnet = false,
    customAssetFilter = () => true,
    excludedChainNames = []
  }: GetCoinGeckoIdByDenomOptions = {}
): string | null {
  const filteredAssetLists = assets.filter(({ chain_name }) => {
    return (
      (allowTestnet || !chain_name.includes('testnet')) &&
      !excludedChainNames.includes(chain_name)
    );
  });

  const filteredAssets = filteredAssetLists
    .flatMap(({ assets }) => assets)
    .filter(({ coingecko_id }) => coingecko_id)
    .filter(customAssetFilter);

  const asset = filteredAssets.find(({ base }) => base === denom);

  return asset?.coingecko_id ?? null;
}

export function getSymbolByChainDenom(
  assets: Asset[],
  denom: CoinDenom
): string {
  const asset = getAssetByDenom(assets, denom);
  const symbol = asset.symbol;
  if (!symbol) {
    return denom;
  }
  return symbol;
}

export function getChainDenomBySymbol(
  assets: Asset[],
  token: string
): CoinDenom {
  const asset = assets.find(({ symbol }) => symbol === token);
  const base = asset?.base;
  if (!base) {
    console.log(`cannot find base for token ${token}`);
    return null;
  }
  return base;
}

export function getExponentByDenom(
  assets: Asset[],
  denom: CoinDenom
): Exponent {
  const asset = getAssetByDenom(assets, denom);
  const unit = asset.denom_units.find(({ denom }) => denom === asset.display);
  return unit?.exponent || 0;
}

export function convertCoinGeckoPricesToDenomPriceMap(
  assets: Asset[],
  prices: Record<string, CoinGeckoUSD>
): PriceHash {
  return Object.keys(prices).reduce((res, geckoId) => {
    const denom = getDenomByCoinGeckoId(assets, geckoId);
    res[denom] = prices[geckoId].usd;
    return res;
  }, {});
}

export function noDecimals(num: number | string): string {
  return new BigNumber(num).decimalPlaces(0, BigNumber.ROUND_DOWN).toString();
}

export function convertBaseUnitsToDollarValue(
  assets: Asset[],
  prices: PriceHash,
  symbol: string,
  amount: string | number
): string {
  const denom = getChainDenomBySymbol(assets, symbol);
  return new BigNumber(amount)
    .shiftedBy(-getExponentByDenom(assets, denom))
    .multipliedBy(prices[denom])
    .toString();
}

export function convertDollarValueToDenomUnits(
  assets: Asset[],
  prices: PriceHash,
  symbol: string,
  value: string | number
): string {
  const denom = getChainDenomBySymbol(assets, symbol);
  return new BigNumber(value)
    .dividedBy(prices[denom])
    .shiftedBy(getExponentByDenom(assets, denom))
    .toString();
}

export function convertBaseUnitsToDisplayUnits(
  assets: Asset[],
  symbol: string,
  amount: string | number
): string {
  const denom = getChainDenomBySymbol(assets, symbol);
  return new BigNumber(amount)
    .shiftedBy(-getExponentByDenom(assets, denom))
    .toString();
}
