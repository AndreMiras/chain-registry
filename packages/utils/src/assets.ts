import { Asset, AssetDenomUnit, AssetList } from '@chain-registry/types';
import BigNumber from 'bignumber.js';

type Denom = AssetDenomUnit['denom'];
type Exponent = AssetDenomUnit['exponent'];

const getAssetByKeyValue = (
  assets: AssetList[],
  key: keyof Asset,
  value: string,
  chainName?: string
): Asset | undefined => {
  const filteredAssets = assets
    .filter(({ chain_name }) => !chainName || chain_name === chainName)
    .flatMap(({ assets }) => assets);

  const matchingAssets = filteredAssets.filter((asset) => asset[key] === value);

  if (matchingAssets.length === 0) {
    return undefined;
  }

  if (matchingAssets.length > 1) {
    throw new Error(
      `Ambiguity: ${matchingAssets.length} assets found for ${key}: ${value}`
    );
  }

  return matchingAssets[0];
};

export const getAssetByDenom = (
  assets: AssetList[],
  denom: Denom,
  chainName?: string
): Asset | undefined => {
  return getAssetByKeyValue(assets, 'base', denom, chainName);
};

export const getAssetBySymbol = (
  assets: AssetList[],
  symbol: string,
  chainName?: string
): Asset | undefined => {
  return getAssetByKeyValue(assets, 'symbol', symbol, chainName);
};

export const getDenomByCoinGeckoId = (
  assets: AssetList[],
  coinGeckoId: string,
  chainName?: string
): Denom | undefined => {
  return getAssetByKeyValue(assets, 'coingecko_id', coinGeckoId, chainName)
    ?.base;
};

type GetCoinGeckoIdByDenomOptions = {
  chainName?: string;
  allowTestnet?: boolean;
  customAssetFilter?: (asset: Asset) => boolean;
  excludedChainNames?: string[];
};

export const getCoinGeckoIdByDenom = (
  assets: AssetList[],
  denom: Denom,
  {
    chainName,
    allowTestnet = false,
    customAssetFilter = () => true,
    excludedChainNames = []
  }: GetCoinGeckoIdByDenomOptions = {}
): string | undefined => {
  const filteredAssetLists = assets.filter(({ chain_name }) => {
    return (
      (!chainName || chain_name === chainName) &&
      (allowTestnet || !chain_name.includes('testnet')) &&
      !excludedChainNames.includes(chain_name)
    );
  });

  const filteredAssets = filteredAssetLists
    .flatMap(({ assets }) => assets)
    .filter(({ coingecko_id }) => coingecko_id)
    .filter(customAssetFilter);

  const asset = filteredAssets.find(({ base }) => base === denom);

  return asset?.coingecko_id;
};

export const getSymbolByDenom = (
  assets: AssetList[],
  denom: Denom,
  chainName?: string
): string | undefined => {
  return getAssetByDenom(assets, denom, chainName)?.symbol;
};

export const getDenomBySymbol = (
  assets: AssetList[],
  symbol: string,
  chainName?: string
): Denom | undefined => {
  return getAssetByKeyValue(assets, 'symbol', symbol, chainName)?.base;
};

export const getExponentFromAsset = (asset: Asset) => {
  return asset.denom_units.find(({ denom }) => denom === asset.display)
    ?.exponent;
};

export const getExponentByDenom = (
  assets: AssetList[],
  denom: Denom,
  chainName?: string
): Exponent | undefined => {
  const asset = getAssetByDenom(assets, denom, chainName);
  return asset ? getExponentFromAsset(asset) : undefined;
};

export const getExponentBySymbol = (
  assets: AssetList[],
  symbol: string,
  chainName?: string
): Exponent | undefined => {
  const asset = getAssetBySymbol(assets, symbol, chainName);
  return asset ? getExponentFromAsset(asset) : undefined;
};

export interface CoinGeckoUSD {
  usd: number;
}

export interface PriceHash {
  [key: Denom]: number;
}

export const convertCoinGeckoPricesToDenomPriceMap = (
  assets: AssetList[],
  prices: Record<string, CoinGeckoUSD>
): PriceHash => {
  return Object.keys(prices).reduce((res: PriceHash, geckoId) => {
    const denom = getDenomByCoinGeckoId(assets, geckoId);
    if (!denom) {
      throw new Error(`No denom found for CoinGecko ID: ${geckoId}`);
    }
    res[denom] = prices[geckoId].usd;
    return res;
  }, {});
};

export const noDecimals = (num: number | string): string => {
  return new BigNumber(num).decimalPlaces(0, BigNumber.ROUND_DOWN).toString();
};

export const convertBaseUnitsToDollarValue = (
  assets: AssetList[],
  prices: PriceHash,
  symbol: string,
  amount: string | number,
  chainName?: string
): string => {
  const asset = getAssetBySymbol(assets, symbol, chainName);
  const denom = asset?.base;
  const exponent = asset ? getExponentFromAsset(asset) : undefined;

  if (!denom || !exponent) {
    throw new Error(`No denom or exponent found for symbol: ${symbol}`);
  }

  return new BigNumber(amount)
    .shiftedBy(-exponent)
    .multipliedBy(prices[denom])
    .toString();
};

export const convertDollarValueToDenomUnits = (
  assets: AssetList[],
  prices: PriceHash,
  symbol: string,
  value: string | number,
  chainName?: string
): string => {
  const asset = getAssetBySymbol(assets, symbol, chainName);
  const denom = asset?.base;
  const exponent = asset ? getExponentFromAsset(asset) : undefined;

  if (!denom || !exponent) {
    throw new Error(`No denom or exponent found for symbol: ${symbol}`);
  }

  return new BigNumber(value)
    .dividedBy(prices[denom])
    .shiftedBy(exponent)
    .toString();
};

export const convertBaseUnitsToDisplayUnits = (
  assets: AssetList[],
  symbol: string,
  amount: string | number,
  chainName?: string
): string => {
  const exponent = getExponentBySymbol(assets, symbol, chainName);

  if (!exponent) {
    throw new Error(`No exponent found for symbol: ${symbol}`);
  }

  return new BigNumber(amount).shiftedBy(-exponent).toString();
};
