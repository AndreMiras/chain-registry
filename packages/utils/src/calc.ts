import { AssetList } from '@chain-registry/types';
import BigNumber from 'bignumber.js';
import {
  Denom,
  getAssetBySymbol,
  getExponentFromAsset,
  getDenomByCoinGeckoId
} from './assets';

export interface CoinGeckoUSDPrice {
  usd: number;
}

export interface DenomPriceMap {
  [key: Denom]: number;
}

export const mapCoinGeckoPricesToDenoms = (
  assets: AssetList[],
  prices: Record<string, CoinGeckoUSDPrice>
): DenomPriceMap => {
  return Object.keys(prices).reduce((res: DenomPriceMap, geckoId) => {
    const denom = getDenomByCoinGeckoId(assets, geckoId);
    if (!denom) {
      throw new Error(`No denom found for CoinGecko ID: ${geckoId}`);
    }
    res[denom] = prices[geckoId].usd;
    return res;
  }, {});
};

export const roundDown = (value: number | string): string => {
  return new BigNumber(value).decimalPlaces(0, BigNumber.ROUND_DOWN).toString();
};

const getAssetInfo = (
  assets: AssetList[],
  symbol: string,
  chainName?: string
) => {
  const asset = getAssetBySymbol(assets, symbol, chainName);
  const denom = asset?.base;
  const exponent = asset ? getExponentFromAsset(asset) : undefined;

  if (!denom || !exponent) {
    throw new Error(`No denom or exponent found for symbol: ${symbol}`);
  }

  return { denom, exponent };
};

const shiftDecimalPlaces = (
  amount: string | number,
  exponent: number,
  direction: 1 | -1 = 1
) => {
  return new BigNumber(amount).shiftedBy(exponent * direction).toString();
};

export const convertBaseUnitToDollarValue = (
  assets: AssetList[],
  prices: DenomPriceMap,
  symbol: string,
  amount: string | number,
  chainName?: string
): string => {
  const { denom, exponent } = getAssetInfo(assets, symbol, chainName);
  const baseAmount = shiftDecimalPlaces(amount, exponent, -1);
  return new BigNumber(baseAmount).multipliedBy(prices[denom]).toString();
};

export const convertDollarValueToBaseUnit = (
  assets: AssetList[],
  prices: DenomPriceMap,
  symbol: string,
  value: string | number,
  chainName?: string
): string => {
  const { denom, exponent } = getAssetInfo(assets, symbol, chainName);
  const baseAmount = new BigNumber(value).dividedBy(prices[denom]).toString();
  return roundDown(shiftDecimalPlaces(baseAmount, exponent));
};

export const convertBaseUnitToDisplayUnit = (
  assets: AssetList[],
  symbol: string,
  amount: string | number,
  chainName?: string
): string => {
  const { exponent } = getAssetInfo(assets, symbol, chainName);
  return shiftDecimalPlaces(amount, exponent, -1);
};

export const convertDisplayUnitToBaseUnit = (
  assets: AssetList[],
  symbol: string,
  amount: string | number,
  chainName?: string
): string => {
  const { exponent } = getAssetInfo(assets, symbol, chainName);
  return roundDown(shiftDecimalPlaces(amount, exponent));
};
