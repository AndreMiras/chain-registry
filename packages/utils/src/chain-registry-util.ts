import { Asset, AssetDenomUnit, AssetList } from '@chain-registry/types';
import BigNumber from 'bignumber.js';

export type Exponent = AssetDenomUnit['exponent'];

export type CoinDenom = AssetDenomUnit['denom'];

export interface CoinGeckoUSD {
  usd: number;
}

export interface PriceHash {
  [key: CoinDenom]: number;
}

export interface ChainRegistryUtilOptions {
  assetLists?: AssetList[];
}

export class ChainRegistryUtil {
  _options?: ChainRegistryUtilOptions = {};
  _asset_lists: AssetList[] = [];

  constructor(options: ChainRegistryUtilOptions = {}) {
    this._options = options;
    this._asset_lists = options.assetLists || [];
  }

  private getAllAssets(chainName: string): Asset[] {
    const targetLists = this._asset_lists.filter(
      (list) => list.chain_name === chainName
    );

    const allAssets: Asset[] = [];
    targetLists.forEach((list) => {
      [].push.apply(allAssets, list.assets);
    });

    return allAssets;
  }

  getAssetByDenom(chainName: string, denom: CoinDenom): Asset {
    const allAssets = this.getAllAssets(chainName);
    const asset = allAssets.find((asset) => asset.base === denom);
    if (!asset) {
      throw new Error(`Asset not found: ${denom}`);
    }
    return asset;
  }

  getDenomByCoinGeckoId(chainName: string, coinGeckoId: string): CoinDenom {
    const allAssets = this.getAllAssets(chainName);
    return allAssets.find((asset) => asset.coingecko_id === coinGeckoId).base;
  }

  chainDenomToSymbol(chainName: string, denom: CoinDenom): string {
    const asset = this.getAssetByDenom(chainName, denom);
    const symbol = asset.symbol;
    if (!symbol) {
      return denom;
    }
    return symbol;
  }

  getSymbolByChainDenom(chainName: string, token: string): CoinDenom {
    const allAssets = this.getAllAssets(chainName);
    const asset = allAssets.find(({ symbol }) => symbol === token);
    const base = asset?.base;
    if (!base) {
      console.log(`cannot find base for token ${token}`);
      return null;
    }
    return base;
  }

  getExponentByDenom(chainName: string, denom: CoinDenom): Exponent {
    const asset = this.getAssetByDenom(chainName, denom);
    const unit = asset.denom_units.find(({ denom }) => denom === asset.display);
    return unit?.exponent || 0;
  }

  convertCoinGeckoPricesToDenomPriceMap(
    chainName: string,
    prices: Record<string, CoinGeckoUSD>
  ): PriceHash {
    return Object.keys(prices).reduce((res, geckoId) => {
      const denom = this.getDenomByCoinGeckoId(chainName, geckoId);
      res[denom] = prices[geckoId].usd;
      return res;
    }, {});
  }

  noDecimals(num: number | string): string {
    return new BigNumber(num).decimalPlaces(0, BigNumber.ROUND_DOWN).toString();
  }

  convertBaseUnitsToDollarValue(
    chainName: string,
    prices: PriceHash,
    symbol: string,
    amount: string | number
  ): string {
    const denom = this.getSymbolByChainDenom(chainName, symbol);
    return new BigNumber(amount)
      .shiftedBy(-this.getExponentByDenom(chainName, denom))
      .multipliedBy(prices[denom])
      .toString();
  }

  convertDollarValueToDenomUnits(
    chainName: string,
    prices: PriceHash,
    symbol: string,
    value: string | number
  ): string {
    const denom = this.getSymbolByChainDenom(chainName, symbol);
    return new BigNumber(value)
      .dividedBy(prices[denom])
      .shiftedBy(this.getExponentByDenom(chainName, denom))
      .toString();
  }

  convertBaseUnitsToDisplayUnits(
    chainName: string,
    symbol: string,
    amount: string | number
  ): string {
    const denom = this.getSymbolByChainDenom(chainName, symbol);
    return new BigNumber(amount)
      .shiftedBy(-this.getExponentByDenom(chainName, denom))
      .toString();
  }
}
