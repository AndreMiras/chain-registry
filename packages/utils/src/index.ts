export * from './chain-registry-util';
export * from './fees';
export * from './utils';
