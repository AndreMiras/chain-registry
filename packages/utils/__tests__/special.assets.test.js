import assets from '../../../__fixtures__/assets.json';
import ibc from '../../../__fixtures__/ibc.json';
import { getAssetLists } from '../src';

it('getAssetLists', () => {
  const chainName = 'persistence';
  const list = getAssetLists(chainName, ibc, assets);
  expect(list).toMatchSnapshot();
});
