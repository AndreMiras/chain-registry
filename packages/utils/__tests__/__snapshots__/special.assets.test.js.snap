// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAssetLists 1`] = `
[
  {
    "assets": [
      {
        "base": "ibc/799FDD409719A1122586A629AE8FCA17380351A51C1F47A80A1B8E7F2A491098",
        "coingecko_id": "akash-network",
        "denom_units": [
          {
            "aliases": [
              "uakt",
            ],
            "denom": "ibc/799FDD409719A1122586A629AE8FCA17380351A51C1F47A80A1B8E7F2A491098",
            "exponent": 0,
          },
          {
            "denom": "akt",
            "exponent": 6,
          },
        ],
        "description": "Akash Token (AKT) is the Akash Network's native utility token, used as the primary means to govern, secure the blockchain, incentivize participants, and provide a default mechanism to store and exchange value.",
        "display": "akt",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/akash/images/akt.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/akash/images/akt.svg",
        },
        "name": "Akash Network",
        "symbol": "AKT",
        "traces": [
          {
            "chain": {
              "channel": "channel-5",
            },
            "counterparty": {
              "chain_name": "akash",
              "channel": "channel-6",
              "denom": "uakt",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/C8A74ABBE2AF892E15680D916A7C22130585CE5704F9B17A10F184A90D53BECA",
        "coingecko_id": "cosmos",
        "denom_units": [
          {
            "aliases": [
              "uatom",
            ],
            "denom": "ibc/C8A74ABBE2AF892E15680D916A7C22130585CE5704F9B17A10F184A90D53BECA",
            "exponent": 0,
          },
          {
            "denom": "atom",
            "exponent": 6,
          },
        ],
        "description": "The native staking and governance token of the Cosmos Hub.",
        "display": "atom",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/cosmoshub/images/atom.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/cosmoshub/images/atom.svg",
        },
        "name": "Cosmos",
        "symbol": "ATOM",
        "traces": [
          {
            "chain": {
              "channel": "channel-24",
            },
            "counterparty": {
              "chain_name": "cosmoshub",
              "channel": "channel-190",
              "denom": "uatom",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/F53B622B0117ECCD0908C1F33ED71B846D9B0C888BE8A16C58098C0AB9497B0A",
        "coingecko_id": "crypto-com-chain",
        "denom_units": [
          {
            "aliases": [
              "basecro",
            ],
            "denom": "ibc/F53B622B0117ECCD0908C1F33ED71B846D9B0C888BE8A16C58098C0AB9497B0A",
            "exponent": 0,
          },
          {
            "denom": "cro",
            "exponent": 8,
          },
        ],
        "description": "CRO coin is the token for the Crypto.com platform.",
        "display": "cro",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/cronos/images/cronos.png",
        },
        "name": "Cronos",
        "symbol": "CRO",
        "traces": [
          {
            "chain": {
              "channel": "channel-11",
            },
            "counterparty": {
              "chain_name": "cryptoorgchain",
              "channel": "channel-17",
              "denom": "basecro",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/5709B8339F05C7E37D27FCFEC18CEDBB0747309144C8B3C4A49C122F0F30B96B",
        "coingecko_id": "iris-network",
        "denom_units": [
          {
            "aliases": [
              "uiris",
            ],
            "denom": "ibc/5709B8339F05C7E37D27FCFEC18CEDBB0747309144C8B3C4A49C122F0F30B96B",
            "exponent": 0,
          },
          {
            "denom": "iris",
            "exponent": 6,
          },
        ],
        "description": "The IRIS token is the native governance token for the IrisNet chain.",
        "display": "iris",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/irisnet/images/iris.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/irisnet/images/iris.svg",
        },
        "name": "IRISnet",
        "symbol": "IRIS",
        "traces": [
          {
            "chain": {
              "channel": "channel-19",
            },
            "counterparty": {
              "chain_name": "irisnet",
              "channel": "channel-15",
              "denom": "uiris",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/CA687302422C075B26948BEC7143F6D47BACAB53421D57DF30741F91D4DFCA1F",
        "coingecko_id": "juno-network",
        "denom_units": [
          {
            "aliases": [
              "ujuno",
            ],
            "denom": "ibc/CA687302422C075B26948BEC7143F6D47BACAB53421D57DF30741F91D4DFCA1F",
            "exponent": 0,
          },
          {
            "denom": "juno",
            "exponent": 6,
          },
        ],
        "description": "The native token of JUNO Chain",
        "display": "juno",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/juno.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/juno.svg",
        },
        "name": "Juno",
        "symbol": "JUNO",
        "traces": [
          {
            "chain": {
              "channel": "channel-37",
            },
            "counterparty": {
              "chain_name": "juno",
              "channel": "channel-33",
              "denom": "ujuno",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/646315E3B0461F5FA4C5C8968A88FC45D4D5D04A45B98F1B8294DD82F386DD85",
        "coingecko_id": "osmosis",
        "denom_units": [
          {
            "aliases": [
              "uosmo",
            ],
            "denom": "ibc/646315E3B0461F5FA4C5C8968A88FC45D4D5D04A45B98F1B8294DD82F386DD85",
            "exponent": 0,
          },
          {
            "aliases": [],
            "denom": "osmo",
            "exponent": 6,
          },
        ],
        "description": "The native token of Osmosis",
        "display": "osmo",
        "keywords": [
          "dex",
          "staking",
        ],
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/osmo.svg",
        },
        "name": "Osmosis",
        "symbol": "OSMO",
        "traces": [
          {
            "chain": {
              "channel": "channel-6",
            },
            "counterparty": {
              "chain_name": "osmosis",
              "channel": "channel-4",
              "denom": "uosmo",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/69B4810E989B56FC6CA41F395228236FA7E7F59B9A3B7AF3FB0A344361C38389",
        "coingecko_id": "ion",
        "denom_units": [
          {
            "aliases": [
              "uion",
            ],
            "denom": "ibc/69B4810E989B56FC6CA41F395228236FA7E7F59B9A3B7AF3FB0A344361C38389",
            "exponent": 0,
          },
          {
            "denom": "ion",
            "exponent": 6,
          },
        ],
        "display": "ion",
        "keywords": [
          "memecoin",
        ],
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/ion.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/osmosis/images/ion.svg",
        },
        "name": "Ion",
        "symbol": "ION",
        "traces": [
          {
            "chain": {
              "channel": "channel-6",
            },
            "counterparty": {
              "chain_name": "osmosis",
              "channel": "channel-4",
              "denom": "uion",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/7D053E5CF515235C2C8A923840D6D5352F19B08722B774355AC97C596C729D5C",
        "coingecko_id": "regen",
        "denom_units": [
          {
            "aliases": [
              "uregen",
            ],
            "denom": "ibc/7D053E5CF515235C2C8A923840D6D5352F19B08722B774355AC97C596C729D5C",
            "exponent": 0,
          },
          {
            "denom": "regen",
            "exponent": 6,
          },
        ],
        "description": "REGEN coin is the token for the Regen Network Platform",
        "display": "regen",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/regen/images/regen.png",
        },
        "name": "Regen Network",
        "symbol": "REGEN",
        "traces": [
          {
            "chain": {
              "channel": "channel-21",
            },
            "counterparty": {
              "chain_name": "regen",
              "channel": "channel-14",
              "denom": "uregen",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/4F03563F9E504B0C147FD395DBCF0A41A45E9A1134DFEC83A94B5E11D43EAA04",
        "coingecko_id": "sentinel",
        "denom_units": [
          {
            "aliases": [
              "udvpn",
            ],
            "denom": "ibc/4F03563F9E504B0C147FD395DBCF0A41A45E9A1134DFEC83A94B5E11D43EAA04",
            "exponent": 0,
          },
          {
            "denom": "dvpn",
            "exponent": 6,
          },
        ],
        "description": "DVPN is the native token of the Sentinel Hub.",
        "display": "dvpn",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/sentinel/images/dvpn.png",
        },
        "name": "Sentinel",
        "symbol": "DVPN",
        "traces": [
          {
            "chain": {
              "channel": "channel-22",
            },
            "counterparty": {
              "chain_name": "sentinel",
              "channel": "channel-15",
              "denom": "udvpn",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/BA51309DEEE69EF9159FADDCD9F566C372157A1F4E559076D201B255C27F484A",
        "coingecko_id": "sifchain",
        "denom_units": [
          {
            "aliases": [
              "rowan",
            ],
            "denom": "ibc/BA51309DEEE69EF9159FADDCD9F566C372157A1F4E559076D201B255C27F484A",
            "exponent": 18,
          },
        ],
        "description": "Rowan Token (ROWAN) is the Sifchain Network's native utility token, used as the primary means to govern, provide liquidity, secure the blockchain, incentivize participants, and provide a default mechanism to store and exchange value.",
        "display": "rowan",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/sifchain/images/rowan.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/sifchain/images/rowan.svg",
        },
        "name": "Sifchain Rowan",
        "symbol": "ROWAN",
        "traces": [
          {
            "chain": {
              "channel": "channel-26",
            },
            "counterparty": {
              "chain_name": "sifchain",
              "channel": "channel-7",
              "denom": "rowan",
            },
            "type": "ibc",
          },
        ],
      },
      {
        "base": "ibc/43F1FDAC4E552DD8194B3FFD35EE16957BC39AFF5CA57545DB71161FA66E7DC0",
        "coingecko_id": "starname",
        "denom_units": [
          {
            "aliases": [
              "uiov",
            ],
            "denom": "ibc/43F1FDAC4E552DD8194B3FFD35EE16957BC39AFF5CA57545DB71161FA66E7DC0",
            "exponent": 0,
          },
          {
            "denom": "iov",
            "exponent": 6,
          },
        ],
        "description": "IOV coin is the token for the Starname (IOV) Asset Name Service",
        "display": "iov",
        "logo_URIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/starname/images/iov.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/starname/images/iov.svg",
        },
        "name": "Starname",
        "symbol": "IOV",
        "traces": [
          {
            "chain": {
              "channel": "channel-27",
            },
            "counterparty": {
              "chain_name": "starname",
              "channel": "channel-13",
              "denom": "uiov",
            },
            "type": "ibc",
          },
        ],
      },
    ],
    "chain_name": "persistence",
  },
]
`;
